### DISCUSSION TITLE
Getting the public root key doesn't work for ACR

### DISCUSSION THREAD

**Discussion by apopaa** [sse-secure-systems/connaisseur#252] [Q&A] [answered]
Method of getting public root key described in [Getting the public root key](https://sse-secure-systems.github.io/connaisseur/v2.1.0/validators/notaryv1/#getting-the-public-root-key) does not work for images in Azure Container Registry:
_> docker run --rm docker.io/securesystemsengineering/get-public-root-key -i apopa/testimage3 -s myregistry.azurecr.io
Traceback (most recent call last):
  File "/app/get_root_key.py", line 54, in <module>
    root_key_id, root_key = asyncio.run(
  File "/usr/local/lib/python3.9/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/usr/local/lib/python3.9/asyncio/base_events.py", line 642, in run_until_complete
    return future.result()
  File "/app/get_root_key.py", line 16, in get_pub_root_key
    root_td = await notary.get_trust_data(image, TUFRole("root"))
  File "/app/connaisseur/validators/notaryv1/notary.py", line 100, in get_trust_data
    raise UnreachableError(
connaisseur.exceptions.UnreachableError: {'message': 'Unable to reach notary host no.', 'context': {'notary_name': 'no', 'tuf_role': 'root', 'detection_mode': False}}_

**Comment by apopaa** [comment, marked as answer]
This otoh works (as well as your example image):
_> docker run --rm docker.io/securesystemsengineering/get-public-root-key -i busybox -s notary.docker.io
KeyID: 074cad59e43e13b440b11d1b5521e20aa8633fc8f3928720590268895711d0c6
Key: -----BEGIN PUBLIC KEY-----
MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEOXYta5TgdCwXTCnLU09W5T4M4r9f
QQrqJuADP6U7g5r9ICgPSmZuRHP/1AYUfOQW3baveKsT969EfELKj1lfCA==
-----END PUBLIC KEY-----_

So Docker Hub works, ACR doesn't work.

**Reply by apopaa** [reply]
FYI, the following REST call works for me for ACR:
_> curl -u "user:password" "https://myregistry.azurecr.io/v2/myregistry.azurecr.io/apopa/testimage3/_trust/tuf/root.json"_

I think the URL that works for ACR is close to what your Python code constructs for Docker Hub, maybe a slight difference that you can either fix (if a bug) or control via a CLI flag (similar to isACR in values.yml).

**Reply by apopaa** [reply]
For anybody else using ACR, this is what I have as a workaround until this issue is fixed:
_curl -u "user:password" "https://myregistry.azurecr.io/v2/myregistry.azurecr.io/apopa/testimage3/\_trust/tuf/root.json" > out.json
--- Get the value of the key signed.keys[root_key_id].keyval.public, where root_key_id = signatures[0].keyid;
--- save this value (without the quotes) in file root_cert_base64.txt
base64 -d root_cert_base64.txt | openssl x509 -pubkey -noout -in - > root.pub_

**Reply by xopham** [reply]
@apopaa thanks for investigating and proposing a workaround :+1: 
The `get-public-root-key` utility does currently not support authentication and I believe this might be causing the problem here. I created an issue for that #254

**Reply by apopaa** [reply]
Yes, I noticed that the underlying library you're using has a "token" parameter that you're not exposing for setting. But that library sets the authorization header to a bearer token, while I used basic authorization with user and password in my curl command. I'm not sure ACR supports bearer token authentication; if it doesn't then the change you have to make is more than exposing the token parameter.

**Reply by xopham** [reply]
Good catch! Thanks for letting us know. We'll see if it works and either expose the token or find another way :slightly_smiling_face:
